### PR TITLE
Show supporter filters for non-supporters

### LINKED ISSUE #3278: Supporter-only search options should show for non-supporters
Options which are available only to supporters should show for non-supporters so they know that they exist.

In the case the user is not a supporter, it should show the "no results found" screen, ie. this layout: 

but the text should be "`this filter requires an active <a>supporter tag</a>`" and the pippi should be the following:

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/58911319-supporter-only-search-options-should-show-for-non-supporters?utm_campaign=plugin&utm_content=tracker%2F21853994&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F21853994&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by char**
Closes #3278 

The final product
First PR 🎉

### CODE REVIEW

File: resources/assets/coffee/react/beatmaps/main.coffee [hunk 1]
```diff
@@ -114,21 +114,32 @@ class Beatmaps.Main extends React.PureComponent
...
           div
             className: 'beatmapsets__content'
-            if @state.beatmaps.length > 0
-              el BeatmapList,
-                items: _.chunk(@state.beatmaps, @state.columnCount)
-                itemBuffer: 5
-                itemHeight: ITEM_HEIGHT
-
-            else
+            if (@state.filters.played != null || @state.filters.rank.length > 0) && !currentUser.is_supporter
```
Comment (nanaya) [inline resources/assets/coffee/react/beatmaps/main.coffee]: should be a function to also be used in `search` function to abort early.

File: resources/lang/en/beatmaps.php [hunk 2]
```diff
@@ -168,6 +168,8 @@
         'search' => [
             'prompt' => 'type in keywords...',
             'options' => 'More Search Options',
+            'supporter-filter' => 'this filter requires supporter',
+            'supporter-filter-quote' => "this filter requires an active <a href='".route('store.products.show', 'supporter-tag')."' target='_blank'>supporter tag</a>",
```
Comment (nanaya) [inline resources/lang/en/beatmaps.php]: Link shouldn't be generated here. See [this](https://github.com/ppy/osu-web/blob/afdd8083a6fc062af5154af4ac6cd818de966c9e/resources/lang/en/store.php#L40) for example.

File: resources/assets/coffee/react/beatmaps/main.coffee [hunk 3]
```diff
@@ -114,21 +114,32 @@ class Beatmaps.Main extends React.PureComponent
...
               div className: 'beatmapsets__empty',
                 el Img2x,
-                  src: '/images/layout/beatmaps/not-found.png'
-                  alt: osu.trans("beatmaps.listing.search.not-found")
-                  title: osu.trans("beatmaps.listing.search.not-found")
-                osu.trans("beatmaps.listing.search.not-found-quote")
+                  src: '/images/layout/beatmaps/supporter-required.png'
+                  alt: osu.trans("beatmaps.listing.search.supporter-filter")
+                  title: osu.trans("beatmaps.listing.search.supporter-filter")
+
```
Comment (notbakaneko) [inline resources/assets/coffee/react/beatmaps/main.coffee]: there's a giant space on this line

File: resources/assets/coffee/react/beatmaps/main.coffee [hunk 4]
```diff
@@ -143,6 +154,7 @@ class Beatmaps.Main extends React.PureComponent
       if value? && BeatmapsetFilter.getDefault(params, key) != value
         charParams[keyToChar[key]] = value
 
+    delete charParams[keyToChar['played']] if !currentUser.is_supporter
```
Comment (nanaya) [inline resources/assets/coffee/react/beatmaps/main.coffee]: this can be removed. Also for `rank` below.

File: resources/lang/en/beatmaps.php [hunk 5]
```diff
@@ -168,6 +168,8 @@
         'search' => [
             'prompt' => 'type in keywords...',
             'options' => 'More Search Options',
+            'supporter-filter' => 'this filter requires supporter',
```
Comment (nanaya) [inline resources/lang/en/beatmaps.php]: change "this filter" to either "this filter combination" or ":filter_names filter" (with `filter_names` being the list of filters).

File: resources/assets/coffee/react/beatmaps/search-panel.coffee [hunk 6]
```diff
@@ -133,24 +133,21 @@ class Beatmaps.SearchPanel extends React.PureComponent
...
-            el Beatmaps.SearchFilter,
-              key: 'played'
-              name: 'played'
-              title: osu.trans('beatmaps.listing.search.filters.played')
-              options: filters.played
-              default: @props.filterDefaults.played
-              selected: @props.filters.played
-          ]
+        el Beatmaps.SearchFilter,
+          key: 'rank'
```
Comment (nanaya) [inline resources/assets/coffee/react/beatmaps/search-panel.coffee]: not needed (also for `played` below).

**Comment by notbakaneko** [comment]
@char are you still working on this?

**Comment by notbakaneko** [comment]
I'm taking over this PR since it's been sitting here for 3 weeks

File: resources/assets/coffee/react/beatmaps/main.coffee [hunk 7]
```diff
@@ -252,6 +273,17 @@ class Beatmaps.Main extends React.PureComponent
...
 
+  supporterFilters: =>
+    _.reject ['played', 'rank'], (name) =>
+      _.isEmpty @state.filters[name]
+
+
+  supporterFiltersTrans: =>
+    _(@supporterFilters()).map (name) ->
+      osu.trans "beatmaps.listing.search.filters.#{name}"
+    .join(', ')
```
Comment (nanaya) [inline resources/assets/coffee/react/beatmaps/main.coffee]: there's this fancy `osu.transArray`

File: resources/lang/en/beatmaps.php [hunk 8]
```diff
@@ -191,6 +192,11 @@
                 'relevance' => 'relevance',
                 'nominations' => 'nominations',
             ],
+            'supporter_filter_quote' => [
+                'before_link' => 'Filtering by :filters requires an active ',
+                'link_text' => 'supporter tag',
+                'after_link' => '',
```
Comment (nanaya) [inline resources/lang/en/beatmaps.php]: crowdin doesn't allow translation of empty text so have fun using that `dangerouslySetInnerHTML`.

File: resources/lang/en/beatmaps.php [hunk 9]
```diff
@@ -168,6 +168,7 @@
         'search' => [
             'prompt' => 'type in keywords...',
             'options' => 'More Search Options',
+            'supporter_filter' => 'Filtering by :filters requires supporter',
```
Comment (nanaya) [inline resources/lang/en/beatmaps.php]: ...tag? Also maybe use same wordings as the other one?

**Pull request merged**